# Logistic Regression

### Its working as scale [ 0 - 0.5 - 1 ]


F(x) = 1 / 1 + e^-z

F(x) = this is point located in line between 0 and 1 included

    F(x)     = probability of class 1 event classification
    1 - F(x) = probability of class 0 event classification


# Log Loss

    The more smaller the log loss value, the bigger chance to get the correct result

logloss = -y * ln(p) - (1 - y) * ln(1 - p)

where y - can be 0 or 1

and p - is the algorithm calculations that this event belong to class 1and p - is the algorithm calculations that this event belongs to class 1

in other words p = F(x)

LogLoss => The point is to get as small as possible different between p and y 

## Working with Log Loss

In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression

Use titanic dataset from kaggle: https://www.kaggle.com/c/titanic/data

In [2]:
data = pd.read_csv('./train.csv')

In [3]:
data.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

Set PassengerId as index

In [4]:
data = data.set_index('PassengerId')

In [5]:
data.head(10)

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   
6                   0       3   
7                   0       1   
8                   0       3   
9                   1       3   
10                  1       2   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   
6                                             Moran, Mr. James    male   NaN   
7                                      McCarthy, Mr. Timothy J    male  54.0   
8                               Palsson, Master. Gosta Leonard    male   2.0   
9            Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0   
10                         Nasser, Mrs. Nicholas (Adele Achem)  female  14.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  
PassengerId                                                          
1                1      0         A/5 21171   7.2500   NaN        S  
2                1      0          PC 17599  71.2833   C85        C  
3                0      0  STON/O2. 3101282   7.9250   NaN        S  
4                1      0            113803  53.1000  C123        S  
5                0      0            373450   8.0500   NaN        S  
6                0      0            330877   8.4583   NaN        Q  
7                0      0             17463  51.8625   E46        S  
8                3      1            349909  21.0750   NaN        S  
9                0      2            347742  11.1333   NaN        S  
10               1      0            237736  30.0708   NaN        C

Check Columns names

In [6]:
data.columns

Index(['Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch', 'Ticket',
       'Fare', 'Cabin', 'Embarked'],
      dtype='object')

### Our mission is to:
### calculate, did the passenger survived or not
### 1 - Yes, 0 - No

In [7]:
target = 'Survived'

In [8]:
# y = 1 - Survived, 0 - did not

y = data[target] 

In [9]:
X = data.drop(target, axis = 1)

In [10]:
X.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 891 entries, 1 to 891
Data columns (total 10 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Pclass    891 non-null    int64  
 1   Name      891 non-null    object 
 2   Sex       891 non-null    object 
 3   Age       714 non-null    float64
 4   SibSp     891 non-null    int64  
 5   Parch     891 non-null    int64  
 6   Ticket    891 non-null    object 
 7   Fare      891 non-null    float64
 8   Cabin     204 non-null    object 
 9   Embarked  889 non-null    object 
dtypes: float64(2), int64(3), object(5)
memory usage: 76.6+ KB


#### Values less than 891 contains empty fields

Example: Cabin, Age, Embarked

Also columns like: Name, Ticket has no value

### Delete Columns that has no Value

In [11]:
# 1 - vertical axis
X = X.drop(
    ['Cabin', 'Name', 'Ticket'], axis = 1
)

### Column 'Age' has empty values
### fill empty values with mean() values

In [12]:
# mean_age = sum(age of all passengers) / passengers qty

mean_age = X['Age'].mean()
mean_age

29.69911764705882

### Replace emptiness with data

In [13]:
# Show all ages in unique view

X['Age'].unique()

array([22.  , 38.  , 26.  , 35.  ,   nan, 54.  ,  2.  , 27.  , 14.  ,
        4.  , 58.  , 20.  , 39.  , 55.  , 31.  , 34.  , 15.  , 28.  ,
        8.  , 19.  , 40.  , 66.  , 42.  , 21.  , 18.  ,  3.  ,  7.  ,
       49.  , 29.  , 65.  , 28.5 ,  5.  , 11.  , 45.  , 17.  , 32.  ,
       16.  , 25.  ,  0.83, 30.  , 33.  , 23.  , 24.  , 46.  , 59.  ,
       71.  , 37.  , 47.  , 14.5 , 70.5 , 32.5 , 12.  ,  9.  , 36.5 ,
       51.  , 55.5 , 40.5 , 44.  ,  1.  , 61.  , 56.  , 50.  , 36.  ,
       45.5 , 20.5 , 62.  , 41.  , 52.  , 63.  , 23.5 ,  0.92, 43.  ,
       60.  , 10.  , 64.  , 13.  , 48.  ,  0.75, 53.  , 57.  , 80.  ,
       70.  , 24.5 ,  6.  ,  0.67, 30.5 ,  0.42, 34.5 , 74.  ])

In [14]:
# This needs to be Fixed

X[(X['Age'].isnull())]['Age']

PassengerId
6     NaN
18    NaN
20    NaN
27    NaN
29    NaN
       ..
860   NaN
864   NaN
869   NaN
879   NaN
889   NaN
Name: Age, Length: 177, dtype: float64

In [15]:
# Lets take those values as Kids Ages

X[(X['Age'] < 1)]['Age']

PassengerId
79     0.83
306    0.92
470    0.75
645    0.75
756    0.67
804    0.42
832    0.83
Name: Age, dtype: float64

In [16]:
X['Age'] = X['Age'].fillna(mean_age)

In [17]:
# Now All NaN = X['Age'].mean()

X['Age'].unique()

array([22.        , 38.        , 26.        , 35.        , 29.69911765,
       54.        ,  2.        , 27.        , 14.        ,  4.        ,
       58.        , 20.        , 39.        , 55.        , 31.        ,
       34.        , 15.        , 28.        ,  8.        , 19.        ,
       40.        , 66.        , 42.        , 21.        , 18.        ,
        3.        ,  7.        , 49.        , 29.        , 65.        ,
       28.5       ,  5.        , 11.        , 45.        , 17.        ,
       32.        , 16.        , 25.        ,  0.83      , 30.        ,
       33.        , 23.        , 24.        , 46.        , 59.        ,
       71.        , 37.        , 47.        , 14.5       , 70.5       ,
       32.5       , 12.        ,  9.        , 36.5       , 51.        ,
       55.5       , 40.5       , 44.        ,  1.        , 61.        ,
       56.        , 50.        , 36.        , 45.5       , 20.5       ,
       62.        , 41.        , 52.        , 63.        , 23.5 

In [18]:
# Lets see what left

X.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 891 entries, 1 to 891
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Pclass    891 non-null    int64  
 1   Sex       891 non-null    object 
 2   Age       891 non-null    float64
 3   SibSp     891 non-null    int64  
 4   Parch     891 non-null    int64  
 5   Fare      891 non-null    float64
 6   Embarked  889 non-null    object 
dtypes: float64(2), int64(3), object(2)
memory usage: 55.7+ KB


#### Embarked = 889, need to fix this
#### Dtype object - this is text field

In [19]:
X['Embarked'].dtype

dtype('O')

In [20]:
X['Embarked'].head()

PassengerId
1    S
2    C
3    S
4    S
5    S
Name: Embarked, dtype: object

In [21]:
# Is there any ampty values?

X.Embarked.isnull().sum()

2

### In this case replace emty fields with most popular data

In [22]:
# Check MODA

X.Embarked.value_counts()

S    644
C    168
Q     77
Name: Embarked, dtype: int64

In [23]:
# S is most popular Char i.e. moda

embarked_mode = X['Embarked'].mode()[0]
embarked_mode

'S'

#### Replace with fillna()

In [24]:
X['Embarked'] = X['Embarked'].fillna(embarked_mode)

In [25]:
# Count 'S' again

X['Embarked'].value_counts()

S    646
C    168
Q     77
Name: Embarked, dtype: int64

In [26]:
# Lets see what left

X.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 891 entries, 1 to 891
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Pclass    891 non-null    int64  
 1   Sex       891 non-null    object 
 2   Age       891 non-null    float64
 3   SibSp     891 non-null    int64  
 4   Parch     891 non-null    int64  
 5   Fare      891 non-null    float64
 6   Embarked  891 non-null    object 
dtypes: float64(2), int64(3), object(2)
memory usage: 55.7+ KB


#### Show DataFrame

In [27]:
X.head(10)

Pclass     Sex        Age  SibSp  Parch     Fare Embarked
PassengerId                                                           
1                 3    male  22.000000      1      0   7.2500        S
2                 1  female  38.000000      1      0  71.2833        C
3                 3  female  26.000000      0      0   7.9250        S
4                 1  female  35.000000      1      0  53.1000        S
5                 3    male  35.000000      0      0   8.0500        S
6                 3    male  29.699118      0      0   8.4583        Q
7                 1    male  54.000000      0      0  51.8625        S
8                 3    male   2.000000      3      1  21.0750        S
9                 3  female  27.000000      0      2  11.1333        S
10                2  female  14.000000      1      0  30.0708        C

After get rid of NaN values, replace all Objects(Strings)

In [28]:
X['Sex'].value_counts()

male      577
female    314
Name: Sex, dtype: int64

### Replace Female and Male With 1 and 0

In [29]:
X['Sex'] = (X['Sex'] == 'female').astype(int)

In [30]:
# Check Result

X['Sex'].value_counts()

0    577
1    314
Name: Sex, dtype: int64

Embarked has more than 2 unique values, to fix this use Dummy-Values

###  This is always a LAST STEP

In [31]:
X['Embarked'].head(10)

PassengerId
1     S
2     C
3     S
4     S
5     S
6     Q
7     S
8     S
9     S
10    C
Name: Embarked, dtype: object

#### For each value of column 'Embarked' will be created new column.

'Embarked_C','Embarked_Q', 'Embarked_S'

In [32]:
X = pd.get_dummies(X)

In [33]:
X.columns

Index(['Pclass', 'Sex', 'Age', 'SibSp', 'Parch', 'Fare', 'Embarked_C',
       'Embarked_Q', 'Embarked_S'],
      dtype='object')

In [34]:
X.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 891 entries, 1 to 891
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Pclass      891 non-null    int64  
 1   Sex         891 non-null    int64  
 2   Age         891 non-null    float64
 3   SibSp       891 non-null    int64  
 4   Parch       891 non-null    int64  
 5   Fare        891 non-null    float64
 6   Embarked_C  891 non-null    uint8  
 7   Embarked_Q  891 non-null    uint8  
 8   Embarked_S  891 non-null    uint8  
dtypes: float64(2), int64(4), uint8(3)
memory usage: 51.3 KB


### Display Result

In [35]:
X.head(10)

Pclass  Sex        Age  SibSp  Parch     Fare  Embarked_C  \
PassengerId                                                              
1                 3    0  22.000000      1      0   7.2500           0   
2                 1    1  38.000000      1      0  71.2833           1   
3                 3    1  26.000000      0      0   7.9250           0   
4                 1    1  35.000000      1      0  53.1000           0   
5                 3    0  35.000000      0      0   8.0500           0   
6                 3    0  29.699118      0      0   8.4583           0   
7                 1    0  54.000000      0      0  51.8625           0   
8                 3    0   2.000000      3      1  21.0750           0   
9                 3    1  27.000000      0      2  11.1333           0   
10                2    1  14.000000      1      0  30.0708           1   

             Embarked_Q  Embarked_S  
PassengerId                          
1                     0           1  
2                     0           0  
3                     0           1  
4                     0           1  
5                     0           1  
6                     1           0  
7                     0           1  
8                     0           1  
9                     0           1  
10                    0           0

### All Columns contains values

#### Working with values

Column 'Pclass' has more than 2 unique values

In [36]:
X.Pclass.value_counts()

3    491
1    216
2    184
Name: Pclass, dtype: int64

In [37]:
X.Pclass.unique()

array([3, 1, 2])

### Create Dummy Column From int64 type column

Change category type

In [38]:
X['Pclass'] = X['Pclass'].astype('category')

Create Dummy Columns

In [39]:
X = pd.get_dummies(X)

Display Result

In [40]:
X.columns

Index(['Sex', 'Age', 'SibSp', 'Parch', 'Fare', 'Embarked_C', 'Embarked_Q',
       'Embarked_S', 'Pclass_1', 'Pclass_2', 'Pclass_3'],
      dtype='object')

In [41]:
X.head(10)

Sex        Age  SibSp  Parch     Fare  Embarked_C  Embarked_Q  \
PassengerId                                                                  
1              0  22.000000      1      0   7.2500           0           0   
2              1  38.000000      1      0  71.2833           1           0   
3              1  26.000000      0      0   7.9250           0           0   
4              1  35.000000      1      0  53.1000           0           0   
5              0  35.000000      0      0   8.0500           0           0   
6              0  29.699118      0      0   8.4583           0           1   
7              0  54.000000      0      0  51.8625           0           0   
8              0   2.000000      3      1  21.0750           0           0   
9              1  27.000000      0      2  11.1333           0           0   
10             1  14.000000      1      0  30.0708           1           0   

             Embarked_S  Pclass_1  Pclass_2  Pclass_3  
PassengerId                                            
1                     1         0         0         1  
2                     0         1         0         0  
3                     1         0         0         1  
4                     1         1         0         0  
5                     1         0         0         1  
6                     0         0         0         1  
7                     1         1         0         0  
8                     1         0         0         1  
9                     1         0         0         1  
10                    0         0         1         0

# Split train.csv file into 2 categories

    1. Training     X_train, y_train
    2. Validation   X_valid, y_valid

In [42]:
# test_size = size of data to use for modeliing

X_train, X_valid, y_train, y_valid = train_test_split(X, y, test_size = 0.25, random_state = 42)

Display Result

In [43]:
X_train.head(10)

Sex        Age  SibSp  Parch      Fare  Embarked_C  Embarked_Q  \
PassengerId                                                                   
299            0  29.699118      0      0   30.5000           0           0   
885            0  25.000000      0      0    7.0500           0           0   
248            1  24.000000      0      2   14.5000           0           0   
479            0  22.000000      0      0    7.5208           0           0   
306            0   0.920000      1      2  151.5500           0           0   
317            1  24.000000      1      0   26.0000           0           0   
3              1  26.000000      0      0    7.9250           0           0   
406            0  34.000000      1      0   21.0000           0           0   
743            1  21.000000      2      2  262.3750           1           0   
261            0  29.699118      0      0    7.7500           0           1   

             Embarked_S  Pclass_1  Pclass_2  Pclass_3  
PassengerId                                            
299                   1         1         0         0  
885                   1         0         0         1  
248                   1         0         1         0  
479                   1         0         0         1  
306                   1         1         0         0  
317                   1         0         1         0  
3                     1         0         0         1  
406                   1         0         1         0  
743                   0         1         0         0  
261                   0         0         0         1

In [44]:
y_train.head(10)

PassengerId
299    1
885    0
248    1
479    0
306    1
317    1
3      1
406    0
743    1
261    0
Name: Survived, dtype: int64

In [45]:
X_valid.head(10)

Sex        Age  SibSp  Parch     Fare  Embarked_C  Embarked_Q  \
PassengerId                                                                  
710            0  29.699118      1      1  15.2458           1           0   
440            0  31.000000      0      0  10.5000           0           0   
841            0  20.000000      0      0   7.9250           0           0   
721            1   6.000000      0      1  33.0000           0           0   
40             1  14.000000      1      0  11.2417           1           0   
291            1  26.000000      0      0  78.8500           0           0   
301            1  29.699118      0      0   7.7500           0           1   
334            0  16.000000      2      0  18.0000           0           0   
209            1  16.000000      0      0   7.7500           0           1   
137            1  19.000000      0      2  26.2833           0           0   

             Embarked_S  Pclass_1  Pclass_2  Pclass_3  
PassengerId                                            
710                   0         0         0         1  
440                   1         0         1         0  
841                   1         0         0         1  
721                   1         0         1         0  
40                    0         0         0         1  
291                   1         1         0         0  
301                   0         0         0         1  
334                   1         0         0         1  
209                   0         0         0         1  
137                   1         1         0         0

In [46]:
y_valid.head(10)

PassengerId
710    1
440    0
841    0
721    1
40     1
291    1
301    1
334    0
209    1
137    1
Name: Survived, dtype: int64

# Build Model

First create basic model,
then create new _more complex_
Compare it with basic

In [47]:
X_train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 668 entries, 299 to 103
Data columns (total 11 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Sex         668 non-null    int64  
 1   Age         668 non-null    float64
 2   SibSp       668 non-null    int64  
 3   Parch       668 non-null    int64  
 4   Fare        668 non-null    float64
 5   Embarked_C  668 non-null    uint8  
 6   Embarked_Q  668 non-null    uint8  
 7   Embarked_S  668 non-null    uint8  
 8   Pclass_1    668 non-null    uint8  
 9   Pclass_2    668 non-null    uint8  
 10  Pclass_3    668 non-null    uint8  
dtypes: float64(2), int64(3), uint8(6)
memory usage: 35.2 KB


All Values are digits, can start building models

LogisticRegression with fit()

In [48]:
lr = LogisticRegression()

In [49]:
lr.fit(X_train, y_train)

LogisticRegression()

Model results store in y_pred

In [50]:
y_pred = lr.predict(X_valid)

In [51]:
y_pred

array([0, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1,
       0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 0, 0, 0, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0,
       1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 1, 1, 0, 1, 0,
       0, 1, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1,
       0, 0, 0, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0, 0,
       0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1, 0,
       1, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0,
       0, 1, 0])

#### metrics for accurancy (correct answears)

In [52]:
from sklearn.metrics import accuracy_score

In [53]:
accuracy_score(y_valid, y_pred)

0.8026905829596412

Compare the result of the train models

In [54]:
y_pred_train = lr.predict(X_train)

In [55]:
accuracy_score(y_train, y_pred_train)

0.8083832335329342

# Create np.array with results

2 columns

Left with 0

Right with 1

0 became if value < o.5...

1 became if value > 0.5...

In [56]:
y_proba = lr.predict_proba(X_valid)

In [57]:
y_proba

array([[0.88803846, 0.11196154],
       [0.7306994 , 0.2693006 ],
       [0.87015652, 0.12984348],
       [0.08991938, 0.91008062],
       [0.25062017, 0.74937983],
       [0.07826309, 0.92173691],
       [0.3344696 , 0.6655304 ],
       [0.90522441, 0.09477559],
       [0.24542478, 0.75457522],
       [0.10340604, 0.89659396],
       [0.69433675, 0.30566325],
       [0.93487211, 0.06512789],
       [0.62484704, 0.37515296],
       [0.8476136 , 0.1523864 ],
       [0.75859307, 0.24140693],
       [0.07898048, 0.92101952],
       [0.72748704, 0.27251296],
       [0.33438492, 0.66561508],
       [0.70247546, 0.29752454],
       [0.70651162, 0.29348838],
       [0.88386345, 0.11613655],
       [0.6428651 , 0.3571349 ],
       [0.39821924, 0.60178076],
       [0.86944941, 0.13055059],
       [0.89833287, 0.10166713],
       [0.92821101, 0.07178899],
       [0.56597441, 0.43402559],
       [0.72832999, 0.27167001],
       [0.9144287 , 0.0855713 ],
       [0.42638413, 0.57361587],
       [0.

# Store data into file

In [59]:
X_train.to_pickle('X_train.pkl')
y_train.to_pickle('y_train.pkl')

X_valid.to_pickle('X_valid.pkl')
y_valid.to_pickle('y_valid.pkl')